api/lib/opentelemetry/trace/propagation/http_trace_context_extractor.rb

```diff
@@ -0,0 +1,56 @@
+# frozen_string_literal: true
+
+# Copyright 2019 OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+module OpenTelemetry
+  module Trace
+    module Propagation
+      # Extracts context from carriers in the W3C Trace Context format
+      class HttpTraceContextExtractor
+        include Context::Propagation::DefaultGetter
+
+        # Returns a new HttpTraceContextExtractor that extracts context using the
+        # specified header keys
+        #
+        # @param [String] traceparent_header_key The traceparent header key used in the carrier
+        # @param [String] tracestate_header_key The tracestate header key used in the carrier
+        # @return [HttpTraceContextExtractor]
+        def initialize(traceparent_header_key: 'traceparent',
+                       tracestate_header_key: 'tracestate')
+          @traceparent_header_key = traceparent_header_key
+          @tracestate_header_key = tracestate_header_key
+        end
+
+        # Extract a remote {Trace::SpanContext} from the supplied carrier.
+        # Invalid headers will result in a new, valid, non-remote {Trace::SpanContext}.
```

Comment:
This feels wrong to me. If I fail to extract the headers for some reason, I most likely want to create a new root span for a trace. This makes it harder to distinguish that case - I have to check if the extracted `SpanContext` is remote, and ignore it if it is non-remote.

Reply:
I agree this looks a little 🐟 y, but I didn't add this behavior in this PR. This is how it's implemented on master. I'll do some research to figure out the backstory.

Reply:
I'm not exactly sure how we ended up with this implementation, but I definitely think it is wrong. I think we should return `SpanContext::INVALID` when parsing fails and everything else should work out as expected. Do you agree?

Reply:
Now that extract returns a context, I think it makes sense to return the original context when parsing fails. I've gone with that in the current round of changes.